File: week06/explore_phila_data/prepare_phl_li_permits.mjs
import gdal from 'gdal-async';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const RAW_DATA_DIR = path.join(__dirname, 'raw_data/');
const PREPARED_DATA_DIR = path.join(__dirname, 'prepared_data/');

const rawFilename = RAW_DATA_DIR + 'phl_li_permits.gpkg';
const preparedFilename = PREPARED_DATA_DIR + 'phl_li_permits.jsonl';

// Load the data from the GeoPackage file
const source = gdal.open(rawFilename);
const layer = source.layers.get(0);
const features = layer.features;

// Write the data to a JSONL file
const f = await fs.open(preparedFilename, 'w');
for (const feature of features) {
  const fieldNames = feature.fields.getNames();
  const fieldEntries = fieldNames.map((name) => [name, feature.fields.get(name)]);
  const geometry = feature.getGeometry();

  const row = Object.fromEntries(fieldEntries);
  row.geog = geometry.isEmpty() ? null : geometry.toJSON();
  await f.write(JSON.stringify(row) + '\n');
}

console.log(`Processed data into ${preparedFilename}`);

File: week06/explore_phila_data/prepare_phl_li_permits.py
import fiona
import json
import math
import pathlib

RAW_DATA_DIR = pathlib.Path(__file__).parent / 'raw_data'
PREPARED_DATA_DIR = pathlib.Path(__file__).parent / 'prepared_data'

raw_filename = RAW_DATA_DIR / 'phl_li_permits.gpkg'
prepared_filename = PREPARED_DATA_DIR / 'phl_li_permits.jsonl'

with (
    # Load the data from the GeoPackage file
    fiona.open(raw_filename, 'r') as source,

    # Write the data to a JSONL file
    open(prepared_filename, 'w') as f
):
    for feature in source:
        row = fiona.model.to_dict(feature.properties)
        row['geog'] = (
            None if any(math.isnan(c) for c in feature.geometry.coordinates)
            else json.dumps(fiona.model.to_dict(feature.geometry))
        )
        f.write(json.dumps(row) + '\n')

print(f'Processed data into {prepared_filename}')
